src/mysqlsh_helpers.py

```diff
@@ -35,84 +61,143 @@ class MySQL:
         """Determine snap common dir for mysqlsh.
 
+        Raises MySQLUpdateConfigurationError if there was an issue configuring the mysql service.
+
         Returns:
             Path to common dir
         """
-        if os.path.exists("/root/snap/mysql-shell/common"):
-            return "/root/snap/mysql-shell/common"
-        return "/tmp"
+        return "/root/snap/mysql-shell/common"
+
+    def update_mysql_configuration(self):
+        """Add a configuration file for mysqld and restart the mysql service."""
+        try:
+            # target file starts with 'z-' so it has priority over the default config file
+            shutil.copyfile("templates/mysqld.cnf", "/etc/mysql/mysql.conf.d/z-custom-mysqld.cnf")
+
+            restart_mysql_command = ["systemctl", "restart", "mysql"]
+            subprocess.check_output(restart_mysql_command, stderr=subprocess.PIPE)
+        except Exception as e:
+            logger.exception(
+                f"Failed to update mysql config for: {self.instance_address}", exc_info=e
+            )
+            raise MySQLUpdateConfigurationError(e.stderr)
 
     def configure_mysql_users(self):
         """Configure the MySQL users for the instance.
 
-        Creates base `root@%` and `clusteradmin@instance_address` user with the
+        Creates base `root@%` and `<server_config>@%` users with the
         appropriate privileges, and reconfigure `root@localhost` user password.
+
         Raises MySQLConfigureMySQLUsersError if the user creation fails.
         """
-        _script = (
+        # SYSTEM_USER and SUPER privileges to revoke from the root users
+        # Reference: https://dev.mysql.com/doc/refman/8.0/en/privileges-provided.html#priv_super
+        privileges_to_revoke = (
+            "SYSTEM_USER",
+            "SYSTEM_VARIABLES_ADMIN",
+            "SUPER",
+            "REPLICATION_SLAVE_ADMIN",
+            "GROUP_REPLICATION_ADMIN",
+            "BINLOG_ADMIN",
+            "SET_USER_ID",
+            "ENCRYPTION_KEY_ADMIN",
+            "VERSION_TOKEN_ADMIN",
+            "CONNECTION_ADMIN",
+        )
+
+        commands = (
             "SET @@SESSION.SQL_LOG_BIN=0;",
-            f"CREATE USER '{self.cluster_admin_user}'@'{self.instance_address}' IDENTIFIED BY '{self.cluster_admin_password}';",
-            f"GRANT ALL ON *.* TO '{self.cluster_admin_user}'@'{self.instance_address}' WITH GRANT OPTION;",
             f"CREATE USER 'root'@'%' IDENTIFIED BY '{self.root_password}';",
             "GRANT ALL ON *.* TO 'root'@'%' WITH GRANT OPTION;",
+            f"CREATE USER '{self.server_config_user}'@'%' IDENTIFIED BY '{self.server_config_password}';",
+            f"GRANT ALL ON *.* TO '{self.server_config_user}'@'%' WITH GRANT OPTION;",
             "UPDATE mysql.user SET authentication_string=null WHERE User='root' and Host='localhost';",
             f"ALTER USER 'root'@'localhost' IDENTIFIED WITH mysql_native_password BY '{self.root_password}';",
```

Comment:
please don't use `mysql_native_password`! it is unsafe, you can read my article https://www.percona.com/blog/2020/06/12/brute-force-mysql-password-from-a-hash/

Reply:
That's on me.

Reply:
@delgod @paulomach i removed usage of `mysql_native_password`. instead (in 37db8b57689305a169c643d92613f15416ef8710), mysql is invoked twice (once to create the `root@%` user with a password, and the second to configure all users using the password for `root@%` when calling `mysql`). please confirm if [this](https://github.com/canonical/mysql-operator/pull/6/files#diff-6e566a7b93efc3f3e7ac0e6d8989256725f876f06af78c45d0ae3815d4bcf82eR189-R213) is a good approach

Reply:
it is ok